design-proposals/klip-50-partition-and-offset-in-ksqldb.md

```diff
@@ -0,0 +1,253 @@
+  by default and must instead be copied into the message value to be accessible)
+* Other semantics, such as whether these metadata will be included in source descriptions
+  or `PRINT TOPIC` commands, or allowed in `INSERT VALUES` statements
+* Compatibility implications for existing ksqlDB queries
+* High-level implementation strategy
+
+## What is not in scope
+
+* [Exposing Kafka message headers in ksqlDB](https://github.com/confluentinc/ksql/issues/1940).
+  We would like to do this in the future but we'll have a separate KLIP for this as it's a
+  larger discussion.
+
+## Value/Return
+
+ksqlDB will expose message partition and offset as pseudocolumns similar to the `ROWTIME`
+pseudocolumn that exists today. Users will be able to select the partition and offset in
+transient queries, persistent queries, and also use the metadata in filters and UDFs.
+(See ["Public APIS"](#public-apis) below for example syntax.)
+
+This will allow users to more easily inspect data and debug ksqlDB queries:
+```
+SELECT *, ROWPARTITION, ROWOFFSET FROM my_stream EMIT CHANGES;
+```
+experiment with processing only a subset of messages:
+```
+SELECT * FROM my_stream WHERE ROWPARTITION = 0 AND ROWOFFSET > 12 EMIT CHANGES;
+```
+and skip select messages ("poison pills") during processing:
+```
+CREATE STREAM my_cleaned_stream AS SELECT * FROM my_stream WHERE NOT (ROWPARTITION = 0 AND ROWOFFSET = 42);
+```
+
+## Public APIS
+
+We will add two new [pseudocolumns](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-14-rowtime-as-pseudocolumn.md)
+to ksqlDB: `ROWPARTITION` (integer) and `ROWOFFSET` (bigint). We choose these names to be 
+consistent with the existing `ROWTIME` pseudocolumn, and to minimize the chance of conflict 
+with users' existing data columns for compatibility reasons. See 
+["Alternatives Considered"](#alternatives-considered) for discussion on possible alternatives.
+
+### Query Semantics
+
+The behavior of `ROWPARTITION` and `ROWOFFSET` in persistent queries and push queries
+will be analogous to those of the [existing `ROWTIME` pseudocolumn](https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-14-rowtime-as-pseudocolumn.md#queries).
+In summary:
+
+* `SELECT *` projections will not include `ROWPARTITION` and `ROWOFFSET`
+* Users may choose to explicitly add `ROWPARTITION` or `ROWOFFSET` in the selection,
+  which will cause ksqlDB to copy the selected pseudocolumn(s) into the output _value_
+* `ROWPARTITION` and `ROWOFFSET` will become reserved column names, and users will not
+  be allowed to use these names either as source column names or as output column aliases.
+  (Existing persistent queries that already use these names will continue to run undisrupted.
+  See ["Compatibility Implications"](#compatibility-implications) for more.)
+
+Consistent with `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will 
```

Comment:
should we discuss `INSERT INTO` and `INSERT VALUES` here as well? (for the former, what happens if you `INSERT INTO` into a stream that selected partition/offset (I don't think there's a problem, you might just produce garbage data) and for the latter it's probably OK to prohibit inserting those fields, unlike `ROWTIME`)

EDIT: I see you discuss this below for `INSERT VALUES` :)